In [1]:
for i in range(5):
    print(i)

0
1
2
3
4


In [2]:
for i in range(1,5,2):
    print(i)

1
3


In [26]:
student = {'name':'hung',
           'degree': 'master'}
for key, value in student.items():
    print(f"Day la {key}:{value}")

Day la name:hung
Day la degree:master


In [28]:
i = 0
while i<10:
    print(i)
    i += 1

0
1
2
3
4
5
6
7
8
9


In [2]:
i = 0
while True:
    print = i
    i += 1
    if(i>10):
        break

W4A1
[SumOfNumBers]
Viết chương trình Python sử dụng một trong 2 loại vòng lặp for hoặc while để
tính tống các số nguyên từ 1 đến n. (n <= 1000)

In [3]:
n = int(input())
tong = 0
if n<=1000:
    for i in range(1, n+1):
        tong += i
    print(tong)
else:
    print("n phai nho hon 1000")

m = n
tong = 0
while 0<=m<=1000:
    tong += m
    m -= 1
print(tong)


15
15


W4A2
[PrimeNumber]
VIết chương trình nhập vào một số tới khi được một số nguyên dương thì thôi.
Kiểm tra số này có phải là số nguyên tố hay không?

In [7]:
n = int(input())
if n<=1:
    print(f"{n} khong phai la so nguyen to")
else:
    huyen = 0
    for i in range(1,n+1):
        if n%i == 0:
            huyen += 1
    if huyen == 2:
        print(f"{n} la so nguyen to")
    else:       
        print(f"{n} khong phai la so nguyen to")

8 khong phai la so nguyen to


In [16]:
import math
n = int(input())
if n<=1:
    print(f"{n} khong phai la so nguyen to")
else:
    for i in range(2,math.isqrt(n) + 1):
        if n%i == 0:
            print(f"{n} la hop so")
            break
    else:
        print(f"{n} la so nguyen to")

234 la hop so


W4A3
[Factorial]
Viết chương trình tính giai thừa của một số nguyên n cho trước (n!=1×2×…×n)
với 0 < n < 100.

In [20]:
n = int(input())
giai_thua = 1

if 0 < n < 100:
    i = 1
    while i <= n:
        giai_thua *= i
        i += 1
    print(f"Giai thua cua {n} la {giai_thua}")
else:
    print("n khong hop le")

Giai thua cua 3 la 6


W4A4
.[CountingDigits]
Viết chương trình nhập vào một số nguyên n và in ra màn hình số chữ số (trừ dấu)
của số đó. Không dùng kiểu dữ liệu string.

In [4]:
n = abs(int(input()))
count = 0
if n == 0:
    count = 1
else:
    while n > 0:
        n //= 10
        count += 1
print(count)

14


W4A5 [MeaningOfLife]
Khi Douglas Adam viết cuốn sách, The Hitchhiker’s Guide to the Galaxy, ông ta đã
thêm vào đó một vài lời nói đùa, và theo năm tháng nó đã trở nên nổi tiếng hơn cả
bản thân cuốn tiểu thuyết: “Câu trả lời cho câu hỏi lớn nhất về sự sống, về vũ trụ và
tất cả mọi thứ là 42.” Một vài người cuồng tín đã tốn hàng năm trời, với mọi cố
gắng nỗ lực để gán ghép ý nghĩa, biểu trưng cho con số 42 và các “sự cố” liên quan
đến nó.
Với mong muốn hái ra tiền từ sư “ám ảnh” điên rồ này, một cuốn sách mới vừa
được xuất bản, 42: Douglas Adams’ Amazingly Accurate Answer to Life, the
Universe and Everything. Cuốn sách được phát hành trùng với kỉ niệm 10 năm ngày
mất của Adams.
“Câu trả lời rất đơn giản”, như Adams đã nói, “Chỉ là một trò đùa. Tôi cần một con
số, một con số be bé, một con số tầm thường, và tôi đã chọn nó. Hai chữ số, bội của
13, một con số hoàn toàn vô nghĩa đối với mấy tay thầy chùa. Tôi ngồi trên bàn làm
việc, ngắm khu vườn và nghĩ rằng số 42 sẽ được việc lắm đây. Tôi gõ nó ra. Và
hết.”
Xuyên suốt lịch sử, các con số được gán cho những ý nghĩa đặc biệt nào đó. Plato
gọi việc nghiên cứu các biểu tượng số học” là level cao nhất của tri thức. Trong khi
Pythagoras tin rằng các con số đều có linh hồn cũng như sức mạnh kì diệu.
Trong khi đó, hàng triệu người hâm hộ Hitchhiker cho đến ngày nay vẫn kiên trì
giải mã những thứ họ cho rằng đó là những động cơ bí mật của Adams.
Hãy viết một chương trình nhận vào một số nguyên dương n và một dãy gồm n số
nguyên. Nếu n số nguyên đó có số 42 thì in ra màn hình dòng chữ "I′ve found the
meaning of life!", ngược lại in ra dòng chữ "It′s a joke!".

In [7]:
n = int(input())
found = False

numbers = list(map(int, input().split()))

for x in numbers:
    if x == 42:
        found = True
        break

if found:
    print("I've found the meaning of life!")
else:
    print("It's a joke!")


I've found the meaning of life!


In [8]:
numbers = list(map(int, input().split()))

if 42 in numbers:
    print("I've found the meaning of life!")
else:
    print("It's a joke!")


I've found the meaning of life!


W4A6
"[SumPrimesInRange]
Nhập hai số nguyên dương a, b (a ≤ b). Hãy tính tổng các số nguyên tố trong đoạn
[a, b]. Mỗi dòng là một bộ test (a b)."

In [2]:
import math

def is_prime(n):
    if n < 2:
        return False
    for i in range(2, int(math.sqrt(n)) + 1):
        if n % i == 0:
            return False
    return True

try:
    line = input().split()
    if len(line) == 2:
        a = int(line[0])
        b = int(line[1])
        sum = 0
    for i in range(a, b+1):
        if is_prime(i):
            sum += i
    print(sum)
except EOFError:
    pass

220


W4A7
"[LargestPrimeFactor]
Nhập vào số nguyên dương n (n ≥ 2). Hãy in ra ước số nguyên tố lớn nhất của n."

In [13]:
import math

n = int(input())
max_prime = -1

while n % 2 == 0:
    max_prime = 2
    n = n // 2

i = 3
while i * i <= n:
    while n % i == 0:
        max_prime = i
        n = n // i
    i += 2

if n > 1:
    max_prime = n

print(max_prime)


43


W4A8
ReverseAddPalindrome]
Nhập số nguyên dương n. Lặp lại phép toán n = n + reverse(n) cho đến khi n là số
đối xứng (palindrome). In ra số bước và giá trị palindrome thu được.

In [ ]:
n_str = input()
n = int(n_str)
steps = 0

while str(n) != str(n)[::-1]:
    reversed_n = int(str(n)[::-1])
    n = n + reversed_n
    steps += 1

print(steps, n)

0 5665


W4A9
[PerfectSquaresDistinctDigits]
Nhập vào số nguyên dương n. In ra tất cả các số chính phương ≤ n có các chữ số
đôi một khác nhau (in theo thứ tự tăng dần, cùng hàng, cách nhau bởi một dấu
cách). Nếu không có, in ra (no number).

In [30]:
n_str = input()
if n_str:
    n = int(n_str)
    results = []
    i = 0
    
    # Tìm các số chính phương <= n
    while True:
        square = i * i
        if square > n:
            break
        
        # Kiểm tra các chữ số đôi một khác nhau
        s = str(square)
        is_distinct = True
        # Duyệt qua từng chữ số để kiểm tra trùng lặp
        for char in s:
            if s.count(char) > 1:
                is_distinct = False
                break
        
        if is_distinct:
            results.append(str(square))
        i += 1

    # In kết quả theo định dạng yêu cầu
    if len(results) == 0:
        print("(no number)")
    else:
        print(" ".join(results))



0 1 4 9 16 25 36 49 64 81 169 196 256 289 324 361 529


W4A10
"[LongestCollatzUpToN]
Nhập vào số nguyên dương n. Tìm số x trong [1, n] có độ dài dãy Collatz lớn nhất
và in ra x cùng độ dài dãy tương ứng (x L). Nếu có nhiều số cùng độ dài, chọn số
nhỏ nhất."

In [16]:
n = int(input())

max_length = 0
best_x = 0

for x in range(1, n + 1):
    current_n = x
    length = 1
    while current_n != 1:
        if current_n % 2 == 0:
            current_n = current_n // 2
        else:
            current_n = 3 * current_n + 1
        length += 1
    if length > max_length:
        max_length = length
        best_x = x

if n > 0:
    print(f"{best_x} {max_length}")

27 112


W4A11
[EvenDivisors]
Viết chương trình đếm số lượng ước số chẵn của số nguyên dương n với n < 10^6.

In [29]:
import math

n = int(input())

count = 0

for i in range(1, int(math.sqrt(n)) + 1):
    if n % i == 0:
        if i % 2 == 0:
            count += 1

        j = n // i
        if j != i:
            if j % 2 == 0:
                count += 1

print(count)



8


W4A12
[Interest]
Một người có tài khoản tiết kiệm ở ngân hàng và gửi vào X đồng với lãi suất là
0.7% mỗi tháng. Viết chương trình tính số tiền sau N tháng người ấy rút được (cả
gốc và lãi, bỏ qua phần lẻ thập phân).

In [ ]:
data = input().split()
x = float(data[0])
n = int(data[1])

tong_tien = x

for i in range(n):
    tong_tien = tong_tien * 1.007
    
print(int(tong_tien))


W4A13
[ClosePairNumber]
Hai số nguyên dương tạo thành một cặp số thân thiết khi chúng tuân theo quy luật:
Số này bằng tổng tất cả các ước của số kia (trừ chính nó) và ngược lại.
Ví dụ 220 và 284 là một cặp số thân thiết
220 ngoài chính nó ra, nó còn có 11 ước số khác là: 1, 2, 4, 5, 10, 11, 20, 44, 55,
110. Tổng của 11 ước này cho chúng ta 284
284 ngoài bản thân nó, còn có 5 ước số khác là: 1, 2, 4, 71, 142. Tổng của chúng
vừa đúng bằng 220
Nhập hai số từ bàn phím và kiểm tra xem chúng có phải là cặp số thân thiết không.
Nếu có, in ra màn hình "true", ngược lại, in ra "false"

In [19]:
m, n = map(int, input().split())
tong_uoc_m = 0
for i in range(1, m):
    if m % i == 0:
        tong_uoc_m += i

tong_uoc_n = 0
for i in range(1, n):
    if n % i == 0:
        tong_uoc_n += i

if tong_uoc_m == n and tong_uoc_n == m:
    print("true")
else:
    print("false")


false


W4A14
[GCD]
Cho hai số m và n là hai số nguyên dương nhập từ bàn phím. Viết chương trình
tìm ước chung lớn nhất của hai số m và n.

In [23]:
m, n = map(int, input().split())

a = abs(m)
b = abs(n)

if a == 0 and b == 0:
    print(0)
else:
    while b != 0:
        a, b = b, a % b
    
    print(a)


30


W4A15
[ChickenAndDog]
"Vừa gà vừa chó,
Bó lại cho tròn,
Ba mươi sáu con,
Một trăm chân chẵn"
Từ bài toán dân gian trên, mở rộng thành chương trình nhận đầu vào là tổng số
con và tổng số chân của gà và chó. Nếu tìm được số phù hợp, in ra số lượng gà và
số lượng chó. Ngược lại, in ra "invalid".

In [ ]:
con, chan = map(int, input().split())
found = False
for i in range(con + 1):
    for j in range(chan + 1):
        if (2 * i + 4 * j) == chan:
            print(i, j)
            found = True
            break
    if found:
        break
if not found:
    print("invalid")

1 16
